### DISCUSSION TITLE
Disable SSH Kitten for specific hosts (Windows)

### DISCUSSION THREAD

**Discussion by Xevion** [kovidgoyal/kitty#6609] [Q&A] [answered]
Hi, I'm having issues with the SSH Kitten for a specific host: a Windows one, my personal desktop.

```
The system cannot find the file specified.
'die' is not recognized as an internal or external command,
operable program or batch file.
'die' is not recognized as an internal or external command,
operable program or batch file.
Shared connection to 192.168.100.1 closed.
```

I can solve this by doing `\ssh` instead of `ssh`, but that's not really a workaround I want to start using habitually.
The configuration options for the SSH kitten seem pretty elaborate, but my tests have not yielded one that works.

I don't care for using any of Kitty's special features - I just want to SSH into my remote host without any special options or arguments. `ssh username@hostname` is all I'm looking for.

I tried a couple variations of the following config, but nothing really works. Maybe I am misunderstanding what these options do?

```
hostname username@hostname
shell_integration disabled
remote_kitty no
```

**Comment by kovidgoyal** [comment]
https://sw.kovidgoyal.net/kitty/kittens/ssh/#opt-kitten-ssh.delegate

**Reply by Xevion** [reply]
I read this configuration line, but I don't know how to use it.

When I put `delegate`, I get
```bash
❯ ssh xevion@hostname
[251 02:38:17.537602] Ignoring invalid config line: delegate
```

When I put `delegate ssh`, I get
```bash
❯ ssh xevion@hostname
[251 02:39:09.875347] Ignoring unknown config key: delegate
```

I attempted to search online for usages of `delegate`, but nothing comes up - not even in this repository. How does this configuration option work?

**Comment by kovidgoyal** [comment]
Put it in ssh.conf

**Reply by Xevion** [reply]
Yes, `~/.config/kitty/ssh.conf` is where I am placing my `delegate` line. It looks like below:
```conf
hostname xevion@<redacted IP>
shell_integration disabled
remote_kitty no
delegate ssh
```

**Reply by kovidgoyal** [reply]
Update kitty

**Reply by Xevion** [reply]
Done, `v0.29.2`, but the problem remains the same. I'm connecting to the exact hostname specified in the config file, but no variation of the `delegate` line has an effect. The errors with the config line being invalid are gone, though.

**Comment by kovidgoyal** [comment]
Works for me with

kitten ssh --kitten=delegate=ssh localhost

Then ps aux | grep ssh

there is no kitten in the output showing that the ssh kitten did indeed
delegate to ssh, as expected.

**Reply by Xevion** [reply]
Weird, I tried that, and it works; connects without using the SSH Kitten.

**Reply by Xevion** [reply]
Is it possible that my config is improperly formatted? In the wrong directory? Bad encoding?
Or that my hostname isn't matching?
The configuration file's format is not well detailed on here, or I didn't find where it is.

**Reply by kovidgoyal** [reply]
Dont see how since you were getting a message about delegate not being recognized till you updated kitty which means the config was being read. And the following also works as expected

```sh
echo "\nhostname localhost\ndelegate ssh" >> ~/.config/kitty/ssh.conf; kitten ssh localhost
```

**Reply by kovidgoyal** [reply]
Leave off the username.

**Reply by kovidgoyal** [reply]
Both in th econfig file and on the command line

**Reply by Xevion** [reply]
Okay, making a little progress.
The following configurations of hostname worked for me (in using `delegate`)
```
hostname *
hostname xevion@*
hostname 192.168.100.1
hostname XX.YY.40.5
```
Even when using a Host alias (defined in `~/.ssh/config`) like `ssh desktop`, it worked just fine.
It doesn't work when the hostname is defined as an IPv4 address like `192.168.100.1` in conjunction with a username.

The following configurations don't work
```
hostname xevion@192.168.100.1
hostname xevion@XX.YY.40.5
```
`XX` and `YY` here obviously are redacted for my privacy.

**Comment by kovidgoyal** [comment]
On Fri, Sep 08, 2023 at 02:28:27AM -0700, Xevion wrote:
> Okay, making a little progress.
> The following configurations of hostname worked for me (in using `delegate`)
> ```
> hostname *
> hostname xevion@*
> ```
> Even when using a Host alias (defined in `~/.ssh/config`) like `ssh desktop`, it worked just fine.
> It doesn't work when the hostname is defined as an IPv4 address like `192.168.100.1`.


Works fine for me with

hostname 127.0.0.1

**Reply by Xevion** [reply]
Reply has been edited; I didn't realize I had to leave off the username from the hostname. I assumed from some of the things the docs said that it supported any style of hostname configuration. I'm guessing defining hostnames like `username@host` is not intended or available?

**Reply by Xevion** [reply]
For my case, technically this problem is resolved, but there are still cases that I don't know how to solve here.
There doesn't seem to be a way to target a username and hostname at the same time in the `ssh.conf` config,
which is an entirely reasonable want.

**Reply by kovidgoyal** [reply, marked as answer]
https://github.com/kovidgoyal/kitty/commit/af4213579c7d3ee48fdb99aadb671c6a28177e4a

**Reply by rabbiveesh** [reply]
is a release coming soon containing this fix?

**Reply by Xevion** [reply]
Just released in 0.30.0, testing and confirmed that things are working how I expected them to. Wildcards, usernames etc. all match properly with the `hostname` directive. Great work!
